pkg/messaging/mqtt/pubsub_test.go

```diff
@@ -0,0 +1,390 @@
+	token = client.Unsubscribe(fmt.Sprintf("%s.%s", topic, subtopic))
+	token.Wait()
+	assert.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+}
+
+//todo: Add client.Publish() -> Token to check if Subscribe() actually worked.
+func TestSubscribe(t *testing.T) {
+	// Creating client to Publish messages to subscribed topic.
+	client, err := newClient(address, "clientIDD", 30*time.Second)
+	assert.Nil(t, err, fmt.Sprintf("got unexpected error: %s", err))
+	t.Cleanup(func() {
+		client.Unsubscribe()
+		client.Disconnect(5)
+	})
+
+	cases := []struct {
+		desc     string
+		topic    string
+		clientID string
+		err      error
+		handler  messaging.MessageHandler
+	}{
+		{
+			desc:     "Subscribe to a topic with an ID",
+			topic:    topic,
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1"},
+		},
+		{
+			desc:     "Subscribe to the same topic with a different ID",
+			topic:    topic,
+			clientID: "clientid2",
+			err:      nil,
+			handler:  handler{false, "clientid2"},
+		},
+		{
+			desc:     "Subscribe to an already subscribed topic with an ID",
+			topic:    topic,
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1"},
+		},
+		{
+			desc:     "Subscribe to a topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s", topic, subtopic),
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1"},
+		},
+		{
+			desc:     "Subscribe to an already subscribed topic with a subtopic with an ID",
+			topic:    fmt.Sprintf("%s.%s", topic, subtopic),
+			clientID: "clientid1",
+			err:      nil,
+			handler:  handler{false, "clientid1"},
+		},
+		{
+			desc:     "Subscribe to an empty topic with an ID",
+			topic:    "",
+			clientID: "clientid1",
+			err:      mqtt_pubsub.ErrEmptyTopic,
+			handler:  handler{false, "clientid1"},
+		},
+		{
+			desc:     "Subscribe to a topic with empty id",
+			topic:    topic,
+			clientID: "",
+			err:      mqtt_pubsub.ErrEmptyID,
+			handler:  handler{false, ""},
+		},
+	}
+	for _, tc := range cases {
+		err = pubsub.Subscribe(tc.clientID, tc.topic, tc.handler)
+		assert.Equal(t, err, tc.err, fmt.Sprintf("%s: expected: %s, but got: %s", tc.desc, err, tc.err))
+
+		if tc.err == nil {
+			token := client.Subscribe(tc.topic, qos, nil)
+			assert.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+
+			expectedMsg := messaging.Message{
+				Publisher: "clientIDD",
+				Channel:   channel,
+				Subtopic:  subtopic,
+				Payload:   data,
+			}
+			token = client.Publish(tc.topic, qos, false, expectedMsg.Payload)
+			assert.Nil(t, token.Error(), fmt.Sprintf("got unexpected error: %s", token.Error()))
+			token.Wait()
+
+			receivedMsg := <-msgChan
+			assert.Equal(t, expectedMsg.Payload, receivedMsg.Payload, fmt.Sprintf("%s: expected %+v got %+v\n", tc.desc, expectedMsg, receivedMsg))
+		}
+	}
+}
+
+//? Should Test Publish() + Subscribe()
```

Comment:
What's with these question marks?

Reply:
custom comments. To make sure I followed all the remarks😅. I'm removing them.